### ISSUE TITLE
document order of workflow stages and interfaces between plugins and the workflow

### ISSUE THREAD

**Issue by djerius** [PerlAlien/Alien-Build#313]
The order of workflow stages is not explicitly delineated.  For example, where does gather occur relative to patch?  What does prefer actually mean?

Prefer is an odd thing, in that it returns data.  This implies there's an interface defining its relationship between it and the workflow (which can modified, as there are plugins which modify prefer behavior  e.g. the `Prefer::GoodVersion` plugin).

So some parts of the workflow can be modified via plugins; which ones?  I'm certainly a little fuzzy about the overall workflow, the points at which modification is permissible, and the interfaces which must be implemented.

So perhaps documentation of 

- the abstract pieces that make up the workflow,
- understanding how plugins fit into the ecosystem
- An ordered list of stages and (if applicable) the logic which transitions between them
- what information is available at each stage
- where modification of the workflow is permissible, and what interfaces have to be implemented

**Comment by plicease** [comment]
I wrote some perlish pseudo code and a flow diagram that I think will be helpful.  This is a work in progress, feedback welcome.  The intent is to eventually include both of these in the PluginAuthor document to help provide context.

```perl
my $probe;
my $override = override();

if($override eq 'system') {

  $probe = probe();

  if($probe ne 'system') {
    die 'system tool or library not found';
  }

}       

elsif($override eq 'default') {
  $probe = probe();

} else { # $override eq 'share'
  # note that in this instance the    
  # probe hook is never called
  $probe = 'share';
}


if($probe eq 'system') {
  gather_system();


} else { # $probe eq 'share' 

  download();
  extract();
  patch();
  build();
  gather_share();

  if(defined &build_ffi) {
    patch_ffi();
    build_ffi();
    gather_ffi();

  }

}


# Download::Negotiate       
sub download {

  my $content = fetch();
  my $file;

  if($content->type eq 'directory') {

    # decode returns a list of candidates
    my @dir = decode($content);

    # prefer filters/sorts candidates
    @dir = prefer(@dir);

    $file = fetch($dir[0]);

  } else { # $content->type eq 'file'
  
    $file = $content;

  }

  $file->save_to_local_file;
} 
```

![Alien-Build flowchart](https://user-images.githubusercontent.com/354009/184399111-7d9b544a-2468-496d-80d4-6341c08fa2ec.png)

The colored block indicate the alienfile blocks these things belong to.  Hooks are indicated by boxes with bars on the side.  Hooks that can be implemented from an alienfile indicate how in blue (I used `[]` to indicate passing in a array reference, most (all?) can also be defined with a code ref `sub {}`).  The diagram doesn't include when required modules are loaded, I felt that would just clutter the diagram, but some text explaining that when you enter the colored block required modules for that phase will be loaded.  One exception is the configure phase which doesn't have any hooks.

**Comment by plicease** [comment]
Noticed a few bugs in the flow diagram already

 - [ ] <What is override> has a `no` option, that should be `system`
 - [ ] <Probe is system> has two `no` options.

**Comment by plicease** [comment]
#48 is related.

**Comment by plicease** [comment]
![Alien-Build flowchart](https://user-images.githubusercontent.com/354009/184442554-84cafaa5-2494-452f-bdbb-adb025996346.png)
minor fixes.

**Comment by plicease** [comment]
- [ ] "What is override?" should read "override result"
 - [ ] "What is probe" should read "probe result"
 - [ ] "Probe is system" should read "probe result" the "yes" branch should read "system" and the "no" branch should read "share"
 - [ ] "Internet" should read "Internet Content"
 - [ ] "File or Directory" should read "Internet Content Type"

**Comment by plicease** [comment]
Generally speaking:

 * any hook can be overridden by a plugin (or from the alienfile), but some hooks have default behavior if none is specified
 * some of the things that you are asking for are already in the PluginAuthor POD (including expectations of what should be passed in and returned to hooks), but I will re-review to see where I can fill in some missing pieces

**Comment by plicease** [comment]
here is a dev version of the new doco:
https://metacpan.org/release/PLICEASE/Alien-Build-2.54_01/view/lib/Alien/Build/Manual/PluginAuthor.pod

 - [ ] The flowchart looks a little too big on a large screen, so I think setting a max width is in order

Otherwise I think this mostly addresses the issues raised in this ticket.

**Comment by plicease** [comment]
- [ ] there is something messed up about the order of this document Pod::Weaver is moving the html section to the end of the DESCRIPTION section :/

**Comment by plicease** [comment]
Okay this is released as 2.56 and I think it is a big improvement, though I am sure that there are further refinements that can be made.  We can open new tickets to address them, so for now I am going to close this one.

**Comment by djerius** [comment]
Thanks!
